pyro/infer/autoguide/guides.py

```diff
@@ -656,44 +656,76 @@ def _loc_scale(self, *args, **kwargs):
+
     """
 
-    def __init__(self, model, hidden_dim=None, init_loc_fn=init_to_median):
-        self.hidden_dim = hidden_dim
-        self.arn = None
+    def __init__(self, model, init_transform_fn, init_loc_fn=init_to_median):
+        self._init_transform_fn = init_transform_fn
+        self.transform = None
         super().__init__(model, init_loc_fn=init_loc_fn)
 
     def get_posterior(self, *args, **kwargs):
         """
         Returns a diagonal Normal posterior distribution transformed by
-        :class:`~pyro.distributions.transforms.iaf.InverseAutoregressiveFlow`.
-        """
-        if self.latent_dim == 1:
-            raise ValueError('latent dim = 1. Consider using AutoDiagonalNormal instead')
-        if self.hidden_dim is None:
-            self.hidden_dim = self.latent_dim
-        if self.arn is None:
-            self.arn = AutoRegressiveNN(self.latent_dim, [self.hidden_dim])
-
-        iaf = transforms.AffineAutoregressive(self.arn)
-        iaf_dist = dist.TransformedDistribution(dist.Normal(0., 1.).expand([self.latent_dim]), [iaf])
-        return iaf_dist
+        a bijective transform.
+        """
+        if self.transform is None:
+            self.transform = self._init_transform_fn(self.latent_dim)
+        flow_dist = dist.TransformedDistribution(dist.Normal(0., 1.).expand([self.latent_dim]), self.transform)
+        return flow_dist
+
+
+class AutoIAFNormal(AutoNormalizingFlow):
+    """
+    This implementation of :class:`AutoContinuous` uses a Diagonal Normal
+    distribution transformed via a :class:`~pyro.distributions.transforms.AffineAutoregressive`
+    to construct a guide over the entire latent space. The guide does not depend on the model's
+    ``*args, **kwargs``.
+
+    Usage::
+
+        guide = AutoIAFNormal(model, hidden_dim=latent_dim)
+        svi = SVI(model, guide, ...)
+
+    :param callable model: a generative model
+    :param int hidden_dim: number of hidden dimensions in the IAF
+    :param callable init_loc_fn: A per-site initialization function.
+        See :ref:`autoguide-initialization` section for available functions.
+    :param int num_transforms: number of :class:`~pyro.distributions.transforms.AffineAutoregressive`
+        transforms to use in sequence.
+    :param init_transform_kwargs: other keyword arguments taken by
+        :func:`~pyro.distributions.transforms.affine_autoregressive`.
+    """
+
+    def __init__(self, model, hidden_dim=None, init_loc_fn=init_to_median, num_transforms=1, **init_transform_kwargs):
+        super().__init__(model,
+                         init_transform_fn=functools.partial(make_transform, affine_autoregressive,
+                                                             repeats=num_transforms,
+                                                             hidden_dims=hidden_dim,
+                                                             **init_transform_kwargs),
+                         init_loc_fn=init_loc_fn)
```

Comment:
one potentially confusing issue for users is that currently these init functions are sort of vacuous and/or have unexpected behavior in the presence of transforms

Reply:
Hmm..I hadn't thought about that. 

@fritzo - are we even using the initialization if we aren't calling `_init_loc()`?

Reply:
@neerajprad `init_loc_fn()` is still used by `InitMessenger` when creating the prototype trace. However if you don't care about the values in that trace, I would suggest hiding this from `AutoNormalizingFlow`.

Reply:
Thanks for explaining, @fritzo. I think, in that case, it probably won't have much of an effect but we should remove it so as to avoid confusion. This will be a breaking change though for `AuotIAFNormal` and will require us to change a few tests where we use the init_fn arg with parametrized autoguide classes, so I would like to do it in a separate PR, if that's okay.

Reply:
@neerajprad I think we can avoid breaking changes if you want:
1. Remove `init_loc_fn` from the new `AutoNormalizingFlow` class (and passing either `init_to_feasible` or `init_to_sample` to `super().__init__()` since the default `init_to_median()` is more expensive and unused).
2. Preserve the `init_loc_fn` kwarg to `AutoIAFNormal` but discard its value?

Reply:
Okay, done. I have added a `FutureWarning` so that we can remove it at some point, and just to warn users that it doesn't have any effect in practice. According to our guidelines, I suppose we shouldn't actually remove it until Pyro 2.0? 🙂